### PR TITLE
"auto" archive staff photographer's images.

### PR DESCRIPTION
**Pull request by akash1810**
"auto" because the workflow is:

- upload image
- image loader will post a message to a (new) SNS topic
- metadata api has a thrall like message consumer, which will detect if an image is staff, and write to dynamo if so then post a metadata-edit message to thrall's SNS topic
- thrall processes the queue as usual, writing to ES.

There is a delay between the last two steps where the UI may not have a filled in star to denote the image has been archived; possible solutions include:

- move `archived` to the root of the document
- server side notifications to broadcast that the SNS message has been processed post upload and so toggle the star state.

### CODE REVIEW

**Comment by kenoir** [comment]
@akash1810 what happens if the message from the metadata-editor is faster to reach ES than the first message from a Thrall?

**Comment by theefer** [comment]
Ah, I didn't realise the image-loader was sending the message. In my mind the thrall would let people know, using a new topic, that it has indexed an image (or done something else). That'd also solve the racing issue.

**Comment by kenoir** [comment]
A sequence diagram would be really handy in understanding what's going on here.

Is this what's going on?

blank flowchart - new page

**Comment by jamesgorrie** [comment]
>  thrall would let people know

This would then be adding logic to thrall i.e. if staff - archive.
It feels as though thrall should remain dumb?

**Comment by akash1810** [comment]
@kenoir re diagram, yes

**Comment by kenoir** [comment]
@jamesgorrie yeah, but also adding the logic the metadata-api seems incongruent too (as it gives it another slightly tangential purpose). Another service perhaps?

**Comment by akash1810** [comment]
> This would then be adding logic to thrall i.e. if staff - archive.

Not necessarily - atm, the Upload Queue processor has this logic. Thrall would just send a `new image` message.

**Comment by jamesgorrie** [comment]
@akash - I was talking about if we let thrall let people know as @theefer thought it worked

**Comment by theefer** [comment]
> This would then be adding logic to thrall i.e. if staff - archive.

No, thrall should just say "hey, here is a new image I just ingested". Other services would still apply the logic.

**Comment by kenoir** [comment]
Maybe like this:

blank flowchart - new page 2

**Comment by theefer** [comment]
What @kenoir said.

**Comment by akash1810** [comment]
Updated to work per @kenoir's diagram.

**Comment by akash1810** [comment]
@theefer @jamesgorrie @kenoir @tsop14 thumbs?

File: metadata-editor/app/lib/DynamoEdits.scala [hunk 1]
```diff
@@ -0,0 +1,27 @@
+package lib
+
+import com.gu.mediaservice.lib.aws.DynamoDB
+import com.gu.mediaservice.model.Edits
+import play.api.libs.json.{Json, JsObject}
+import play.api.libs.concurrent.Execution.Implicits._
+
+object DynamoEdits {
```
Comment (theefer) [inline metadata-editor/app/lib/DynamoEdits.scala]: Can we have a comment about what this is and why it's here (vs the controller), as it looks transitional (for a reason)?

File: metadata-editor/app/lib/DynamoEdits.scala [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+      .map(publish(id))
+  }
+
+  def publish(id: String)(metadata: JsObject): Edits = {
+    val message = Json.obj(
+      "id" -> id,
+      "data" -> metadata
+    )
+
+    Notifications.publish(message, "update-image-user-metadata")
```
Comment (theefer) [inline metadata-editor/app/lib/DynamoEdits.scala]: Didn't we just change this to be type-safer, or has this not been merged? cc @jamesgorrie

File: metadata-editor/app/lib/MetadataMessageConsumer.scala [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+  Config.queueUrl, Config.awsEndpoint, Config.awsCredentials, MetadataEditorMetrics.processingLatency) {
+
+  override def chooseProcessor(subject: String): Option[JsValue => Future[Any]] =
+    PartialFunction.condOpt(subject) {
+      case "image-indexed" => processIndexedImage
+    }
+
+  def isStaffPhotographer(image: Image) = {
+    image.metadata.byline match {
+      case Some(byline: String) => StaffPhotographers.store.contains(byline)
```
Comment (theefer) [inline metadata-editor/app/lib/MetadataMessageConsumer.scala]: Do we only archive staff photographer's images (not contributed/commissioned)?

File: metadata-editor/app/lib/DynamoEdits.scala (hunk 2 above)
Comment (jamesgorrie) [inline metadata-editor/app/lib/DynamoEdits.scala]: Just merged that change as there was some things to think about...

**Comment by theefer** [comment]
Nice!

A few comments/questions. Would be good to trial on TEST first?

Also note that the branch needs master merged in.

File: metadata-editor/app/lib/MetadataMessageConsumer.scala (hunk 3 above)
Comment (jamesgorrie) [inline metadata-editor/app/lib/MetadataMessageConsumer.scala]: We should do all of them.

**Comment by jamesgorrie** [comment]
Something to think about - how do we do this retrospectively.

**Comment by theefer** [comment]
> how do we do this retrospectively

The reindex message should be sent if we send a media-api reindex (@akash1810 is it?). So we could just query by name or category, trigger a reindex and that should reset the overrides?

Still not 100% this is the right solution, even though I suggested it.

@kenoir @jamesgorrie @akash1810 thoughts welcome.

**Comment by akash1810** [comment]
The message is send when thrall processes an `image` message iirc that message is only sent on new upload.

> how do we do this retrospectively.

I'm not sure this is a concern, as this PR is automating a process that users have been spending time doing manually already.

**Comment by theefer** [comment]
We should probably also send the message on [`update-image`](https://github.com/guardian/grid/blob/master/media-api/app/controllers/MediaApi.scala#L183)? Thoughts?

**Comment by akash1810** [comment]
> We should probably also send the message on update-image? Thoughts?

end of week tiredness kicking in....Thrall processes `image` and `update-image` in the [same way](https://github.com/guardian/grid/blob/master/thrall/app/lib/MessageConsumer.scala#L76)... so we're already sending the message on `update-image` :smile:

File: metadata-editor/app/lib/MetadataMessageConsumer.scala (hunk 3 above)
Comment (akash1810) [inline metadata-editor/app/lib/MetadataMessageConsumer.scala]: any hints as to where these are listed?

File: metadata-editor/app/lib/MetadataMessageConsumer.scala [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+object MetadataMessageConsumer extends MessageConsumer(
+  Config.queueUrl, Config.awsEndpoint, Config.awsCredentials, MetadataEditorMetrics.processingLatency) {
+
+  override def chooseProcessor(subject: String): Option[JsValue => Future[Any]] =
+    PartialFunction.condOpt(subject) {
+      case "image-indexed" => processIndexedImage
+    }
+
+  def isArchivingCategory(category: String) = {
+    val archiveCategory = List("staff-photographer", "contract-photographer", "commissioned-photographer")
```
Comment (akash1810) [inline metadata-editor/app/lib/MetadataMessageConsumer.scala]: magic strings are magic...

Comment (jamesgorrie) [inline metadata-editor/app/lib/MetadataMessageConsumer.scala]: You could match on the types.
```Scala
isArchivingUsageRights[T <: UsageRights](usageRights: T) = usageRights match {
  case _:StaffPhotographer | _:ContractPhotographer | _:CommissionedPhotographer => true
  case _ => false
}
```

(there might be a better syntax, but it's nicer than using the strings).

Comment (akash1810) [inline metadata-editor/app/lib/MetadataMessageConsumer.scala]: Much nicer!

The `case _:T1 | _:T2 => foo()` syntax looks to be pretty common and readable too.

Comment (akash1810) [inline metadata-editor/app/lib/MetadataMessageConsumer.scala]: Much nicer!

The `case _:T1 | _:T2 => foo()` syntax looks to be pretty common and readable too.

**Comment by akash1810** [comment]
One thing to note: manual updates to the byline will not trigger an auto archive atm.

**Comment by jamesgorrie** [comment]
@akash - after talking to @theefer I'm not sure that we should do things like that as it's a little hard for a user to reason with. Similar to the usage rights - we might prompt a user, after setting a staff-photographer, would they like to copy the data to the usage rights from the client, but not automatically.

Similar to this, we could suggest archiving it.

**Comment by akash1810** [comment]
> we might prompt a user, after setting a staff-photographer, would they like to copy the data to the usage rights from the client, but not automatically.

+1

**Comment by theefer** [comment]
Sounds like it might be worth having another catch up about this to work
out the different scenarios?

Note that a lot of this work is likely to be useful regardless, for
instance to clear up entries in dynamo for images after they are deleted.

On Tue, 11 Aug 2015 10:19 James Gorrie <notifications@github.com> wrote:

> @akash <https://github.com/akash> - after talking to @theefer
> <https://github.com/theefer> I'm not sure that we should do things like
> that as it's a little hard for a user to reason with. Similar to the usage
> rights - we might prompt a user, after setting a staff-photographer, would
> they like to copy the data to the usage rights from the client, but not
> automatically.
>
> Similar to this, we could suggest archiving it.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/guardian/grid/pull/1027#issuecomment-129786273>.
>

-- 

------------------------------
Visit theguardian.com. On your mobile and tablet, download the Guardian 
iPhone and Android apps theguardian.com/guardianapp and our tablet editions 
theguardian.com/editions.  Save up to 57% by subscribing to the Guardian 
and Observer - choose the papers you want and get full digital access. 
 Visit subscribe.theguardian.com

This e-mail and all attachments are confidential and may also be 
privileged. If you are not the named recipient, please notify the sender 
and delete the e-mail and all attachments immediately. Do not disclose the 
contents to another person. You may not use the information for any 
purpose, or store, or copy, it in any way.  Guardian News & Media Limited 
is not liable for any computer viruses or other material transmitted with 
or as part of this e-mail. You should employ virus checking software.
 
Guardian News & Media Limited is a member of Guardian Media Group plc. Registered 
Office: PO Box 68164, Kings Place, 90 York Way, London, N1P 2AP.  Registered 
in England Number 908396

**Comment by akash1810** [comment]
(this comment is more of a reminder for me for tomorrow...)

Where do we stand on this?

**Comment by theefer** [comment]
Let's catch up tomorrow

**Comment by jamesgorrie** [comment]
@theefer is it tomorrow yet?

**Comment by theefer** [comment]
Let's try and find some time later today.

**Comment by akash1810** [comment]
Caught up w/master now.

Note: GH seems to show a `git mv` MessageConsumer.scala -> ThrallMessageConsumer.scala as a delete and new file :cry:

**Comment by akash1810** [comment]
This work is superseded by https://github.com/guardian/grid/pull/1202.

**Pull request closed without merging**